File: java/src/main/java/com/adaptionsoft/games/uglytrivia/Game.java
package com.adaptionsoft.games.uglytrivia;

import java.io.PrintStream;
import java.util.ArrayList;
import java.util.LinkedList;

public class Game {
    public static final int NUMBER_OF_PLAYERS = 6;
    public static final int NUMBER_OF_PURSES = 6;
    public static final int NUMBER_OF_PENALTY_BOXES = 6;
    public static final String POP_QUESTION = "Pop Question ";
    public static final String SCIENCE_QUESTION = "Science Question ";
    public static final String SPORTS_QUESTION = "Sports Question ";
    public static final String ROCK_QUESTION = "Rock Question ";
    public static final int QUESTIONS_COUNT_IN_EACH_CATEGORY = 50;
    public static final int MINIMAL_NUMBER_OF_PLAYERS = 2;
    public static final String POP = "Pop";
    public static final String SCIENCE = "Science";
    public static final String SPORTS = "Sports";
    public static final String ROCK = "Rock";
    public static final int FIRST_PLAYER = 0;
    public static final boolean GAME_NOT_OVER = true;
    ArrayList players = new ArrayList();

    public int[] places = new int[NUMBER_OF_PLAYERS];
    int[] purses = new int[NUMBER_OF_PURSES];
    public boolean[] inPenaltyBox = new boolean[NUMBER_OF_PENALTY_BOXES];
    LinkedList popQuestions = new LinkedList();

    LinkedList scienceQuestions = new LinkedList();
    LinkedList sportsQuestions = new LinkedList();
    LinkedList rockQuestions = new LinkedList();
    int currentPlayer = 0;

    public boolean isGettingOutOfPenaltyBox;

    private final PrintStream outputPrintStream;

    public Game(PrintStream outputPrintStream) {
        for (int i = 0; i < QUESTIONS_COUNT_IN_EACH_CATEGORY; i++) {
            popQuestions.addLast(POP_QUESTION + i);
            scienceQuestions.addLast((SCIENCE_QUESTION + i));
            sportsQuestions.addLast((SPORTS_QUESTION + i));
            rockQuestions.addLast(createRockQuestion(i));
        }
        this.outputPrintStream = outputPrintStream;
    }

    public String createRockQuestion(int index) {
        return ROCK_QUESTION + index;
    }

    public boolean isPlayable() {
        return (howManyPlayers() >= MINIMAL_NUMBER_OF_PLAYERS);
    }

    public boolean add(String playerName) {
        players.add(playerName);
        places[howManyPlayers()] = 0;
        purses[howManyPlayers()] = 0;
        inPenaltyBox[howManyPlayers()] = false;

        outputPrintStream.println(playerName + " was added");
        outputPrintStream.println("They are player number " + players.size());
        return true;
    }

    public int howManyPlayers() {
        return players.size();
    }

    public void roll(int roll) {
        outputPrintStream.println(players.get(currentPlayer) + " is the current player");
        outputPrintStream.println("They have rolled a " + roll);

        boolean playerIsInPenaltyBox = inPenaltyBox[currentPlayer];
        if (playerIsInPenaltyBox)
            startTurnInPenaltyBox(roll);
        else
            startNormalTurn(roll);
    }

    private void startTurnInPenaltyBox(int roll) {
        boolean getOutOfPenaltyBox = (roll % 2 != 0);
        if (getOutOfPenaltyBox) {
            isGettingOutOfPenaltyBox = true;

            outputPrintStream.println(players.get(currentPlayer) + " is getting out of the penalty box");
            startNormalTurn(roll);
        } else {
            outputPrintStream.println(players.get(currentPlayer) + " is not getting out of the penalty box");
            isGettingOutOfPenaltyBox = false;
        }
    }

    private void startNormalTurn(int roll) {
        places[currentPlayer] = places[currentPlayer] + roll;
        if (places[currentPlayer] > 11)
            places[currentPlayer] = places[currentPlayer] - 12;

        outputPrintStream.println(players.get(currentPlayer)
                + "'s new location is "
                + places[currentPlayer]);
        outputPrintStream.println("The category is " + currentCategory());
        askQuestion();
    }

    private void askQuestion() {
        if (currentCategory() == POP)
            outputPrintStream.println(popQuestions.removeFirst());
        if (currentCategory() == SCIENCE)
            outputPrintStream.println(scienceQuestions.removeFirst());
        if (currentCategory() == SPORTS)
            outputPrintStream.println(sportsQuestions.removeFirst());
        if (currentCategory() == ROCK)
            outputPrintStream.println(rockQuestions.removeFirst());
    }


    private String currentCategory() {
        if (places[currentPlayer] == 0) return POP;
        if (places[currentPlayer] == 4) return POP;
        if (places[currentPlayer] == 8) return POP;
        if (places[currentPlayer] == 1) return SCIENCE;
        if (places[currentPlayer] == 5) return SCIENCE;
        if (places[currentPlayer] == 9) return SCIENCE;
        if (places[currentPlayer] == 2) return SPORTS;
        if (places[currentPlayer] == 6) return SPORTS;
        if (places[currentPlayer] == 10) return SPORTS;
        return ROCK;
    }

    public boolean wasCorrectlyAnswered() {
        if (inPenaltyBox[currentPlayer]) {
            if (isGettingOutOfPenaltyBox) {
                outputPrintStream.println("Answer was correct!!!!");
                purses[currentPlayer]++;
                outputPrintStream.println(players.get(currentPlayer)
                        + " now has "
                        + purses[currentPlayer]
                        + " Gold Coins.");

                boolean gameIsOver = didPlayerWin();
                currentPlayer++;
                if (currentPlayer == players.size()) currentPlayer = FIRST_PLAYER;

                return gameIsOver;
            } else {
                currentPlayer++;
                if (currentPlayer == players.size()) currentPlayer = FIRST_PLAYER;
                return GAME_NOT_OVER;
            }


        } else {

            outputPrintStream.println("Answer was corrent!!!!");
            purses[currentPlayer]++;
            outputPrintStream.println(players.get(currentPlayer)
                    + " now has "
                    + purses[currentPlayer]
                    + " Gold Coins.");

            boolean winner = didPlayerWin();
            currentPlayer++;
            if (currentPlayer == players.size()) currentPlayer = FIRST_PLAYER;

            return winner;
        }
    }

    public boolean wrongAnswer() {
        outputPrintStream.println("Question was incorrectly answered");
        outputPrintStream.println(players.get(currentPlayer) + " was sent to the penalty box");
        inPenaltyBox[currentPlayer] = true;

        currentPlayer++;
        if (currentPlayer == players.size()) currentPlayer = FIRST_PLAYER;
        return true;
    }


    private boolean didPlayerWin() {
        return !(purses[currentPlayer] == 6);
    }
}

File: kotlin/src/main/kotlin/com/adaptionsoft/games/uglytrivia/Game.kt
package com.adaptionsoft.games.uglytrivia

import java.util.*

class Game {
    var players = ArrayList<String>()
    var places = IntArray(6)
    var purses = IntArray(6)
    var inPenaltyBox = BooleanArray(6)
    var popQuestions = LinkedList<String>()
    var scienceQuestions = LinkedList<String>()
    var sportsQuestions = LinkedList<String>()
    var rockQuestions = LinkedList<String>()
    var currentPlayer = 0
    var isGettingOutOfPenaltyBox: Boolean = false

    init {
        for (i in 0..49) {
            popQuestions.addLast("Pop Question " + i)
            scienceQuestions.addLast("Science Question " + i)
            sportsQuestions.addLast("Sports Question " + i)
            rockQuestions.addLast(createRockQuestion(i))
        }
    }

    fun createRockQuestion(index: Int): String {
        return "Rock Question " + index
    }

    val isPlayable: Boolean
        get() = howManyPlayers() >= 2

    fun add(playerName: String): Boolean {


        players.add(playerName)
        places[howManyPlayers()] = 0
        purses[howManyPlayers()] = 0
        inPenaltyBox[howManyPlayers()] = false

        System.out.println(playerName + " was added")
        System.out.println("They are player number " + players.size)
        return true
    }

    fun howManyPlayers(): Int {
        return players.size
    }

    fun roll(roll: Int) {
        System.out.println(players.get(currentPlayer) + " is the current player")
        System.out.println("They have rolled a " + roll)

        if (inPenaltyBox[currentPlayer]) {
            if (roll % 2 != 0) {
                isGettingOutOfPenaltyBox = true

                System.out.println(players.get(currentPlayer) + " is getting out of the penalty box")
                places[currentPlayer] = places[currentPlayer] + roll
                if (places[currentPlayer] > 11) places[currentPlayer] = places[currentPlayer] - 12

                System.out.println(players.get(currentPlayer)
                        + "'s new location is "
                        + places[currentPlayer])
                System.out.println("The category is " + currentCategory())
                askQuestion()
            } else {
                System.out.println(players.get(currentPlayer) + " is not getting out of the penalty box")
                isGettingOutOfPenaltyBox = false
            }

        } else {

            places[currentPlayer] = places[currentPlayer] + roll
            if (places[currentPlayer] > 11) places[currentPlayer] = places[currentPlayer] - 12

            System.out.println(players.get(currentPlayer)
                    + "'s new location is "
                    + places[currentPlayer])
            System.out.println("The category is " + currentCategory())
            askQuestion()
        }

    }

    private fun askQuestion() {
        if (currentCategory() === "Pop")
            System.out.println(popQuestions.removeFirst())
        if (currentCategory() === "Science")
            System.out.println(scienceQuestions.removeFirst())
        if (currentCategory() === "Sports")
            System.out.println(sportsQuestions.removeFirst())
        if (currentCategory() === "Rock")
            System.out.println(rockQuestions.removeFirst())
    }


    private fun currentCategory(): String {
        if (places[currentPlayer] == 0) return "Pop"
        if (places[currentPlayer] == 4) return "Pop"
        if (places[currentPlayer] == 8) return "Pop"
        if (places[currentPlayer] == 1) return "Science"
        if (places[currentPlayer] == 5) return "Science"
        if (places[currentPlayer] == 9) return "Science"
        if (places[currentPlayer] == 2) return "Sports"
        if (places[currentPlayer] == 6) return "Sports"
        if (places[currentPlayer] == 10) return "Sports"
        return "Rock"
    }

    fun wasCorrectlyAnswered(): Boolean {
        if (inPenaltyBox[currentPlayer]) {
            if (isGettingOutOfPenaltyBox) {
                System.out.println("Answer was correct!!!!")
                purses[currentPlayer]++
                System.out.println(players.get(currentPlayer)
                        + " now has "
                        + purses[currentPlayer]
                        + " Gold Coins.")

                val winner = didPlayerWin()
                currentPlayer++
                if (currentPlayer == players.size) currentPlayer = 0

                return winner
            } else {
                currentPlayer++
                if (currentPlayer == players.size) currentPlayer = 0
                return true
            }


        } else {

            System.out.println("Answer was corrent!!!!")
            purses[currentPlayer]++
            System.out.println(players.get(currentPlayer)
                    + " now has "
                    + purses[currentPlayer]
                    + " Gold Coins.")

            val winner = didPlayerWin()
            currentPlayer++
            if (currentPlayer == players.size) currentPlayer = 0

            return winner
        }
    }

    fun wrongAnswer(): Boolean {
        System.out.println("Question was incorrectly answered")
        System.out.println(players.get(currentPlayer) + " was sent to the penalty box")
        inPenaltyBox[currentPlayer] = true

        currentPlayer++
        if (currentPlayer == players.size) currentPlayer = 0
        return true
    }


    private fun didPlayerWin(): Boolean {
        return purses[currentPlayer] != 6
    }
}
